### PR TITLE
validator: Add support for compressed YAML

### PR DESCRIPTION
**Pull request by sgallagher**
Fixes https://github.com/fedora-modularity/libmodulemd/issues/479

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

**Comment by sgallagher** [comment]
> This looks like a faithful reproduction of the compression handling code found in `modulemd_module_index_update_from_file()` in `modulemd-module-index.c`.
> 
> But I must be missing something. Why can't `parse_file()` be reduced to a small wrapper that simply calls `modulemd_module_index_update_from_file()`?

It needs to be able to call [`modulemd_index_update_from_parser()`](https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.ModuleIndex-(Private).html#modulemd-module-index-update-from-parser) because that function takes one extra argument: `autogen_module_name`. Without this argument, it can't parse `document: modulemd` that doesn't have the module and stream names set (such as those you would find in dist-git).

**Comment by sgallagher** [comment]
> It needs to be able to call [`modulemd_index_update_from_parser()`](https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.ModuleIndex-(Private).html#modulemd-module-index-update-from-parser) because that function takes one extra argument: `autogen_module_name`. Without this argument, it can't parse `document: modulemd` that doesn't have the module and stream names set (such as those you would find in dist-git).

I suppose an alternative approach would be to create a private `modulemd_update_from_file_ext()` function that both could use. That would probably be better, as it would avoid the code duplication. I will just do that.

**Comment by mmathesius** [comment]
> > It needs to be able to call [`modulemd_index_update_from_parser()`](https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.ModuleIndex-(Private).html#modulemd-module-index-update-from-parser) because that function takes one extra argument: `autogen_module_name`. Without this argument, it can't parse `document: modulemd` that doesn't have the module and stream names set (such as those you would find in dist-git).

_That's_ what I was missing. So close...

> I suppose an alternative approach would be to create a private `modulemd_update_from_file_ext()` function that both could use. That would probably be better, as it would avoid the code duplication. I will just do that.

I'd be happy to take this on--unless you really want to (or almost have it done already).

**Comment by sgallagher** [comment]
> > > It needs to be able to call [`modulemd_index_update_from_parser()`](https://fedora-modularity.github.io/libmodulemd/latest/modulemd-2.0-Modulemd.ModuleIndex-(Private).html#modulemd-module-index-update-from-parser) because that function takes one extra argument: `autogen_module_name`. Without this argument, it can't parse `document: modulemd` that doesn't have the module and stream names set (such as those you would find in dist-git).
> 
> _That's_ what I was missing. So close...
> 
> > I suppose an alternative approach would be to create a private `modulemd_update_from_file_ext()` function that both could use. That would probably be better, as it would avoid the code duplication. I will just do that.
> 
> I'd be happy to take this on--unless you really want to (or almost have it done already).

Thanks for the offer, but I just did it. See the new version I just pushed.

File: modulemd/modulemd-validator.c [hunk 1]
```diff
@@ -114,36 +114,16 @@ parse_file (const gchar *filename, GPtrArray **failures, GError **error)
   MMD_INIT_YAML_PARSER (parser);
   MMD_INIT_YAML_EVENT (event);
   g_autoptr (FILE) yaml_stream = NULL;
```
Comment (mmathesius) [inline modulemd/modulemd-validator.c:116]:
```suggestion
```
Now extraneous.

**Pull request merged**